### PR TITLE
Reduce --long mode MT jobsize at higher levels

### PR DESCRIPTION
**Pull request by senhuang42**
Brought up in #2352. Since `cycleLog(chainLog)` is just `chainLog-1` if the compression strategy is `btlazy2` or higher. The effect of this change is that we use smaller jobsizes at higher compression levels with `--long`, allowing greater use of multithreading on files a few hundred megabytes large.

This PR also corrects an error in `FIO_adjustParamsForPatchFromMode()` related to `cycleLog()`.

| Conf        | dev, no `--long` | dev, `--long`: `MAX(21, chainLog+4)` | change_ldm_config, `--long`: `MAX(21, ZSTD_cycleLog(chainLog, strategy)+4` |
|-------------|------------------|--------------------------------------|----------------------------------------------------------------------|
| -22 wlog=27 | 29               | 30                                   | 30                                                                   |
| -19 wlog=27 | 29               | 28                                   | 27                                                                   |
| -19 wlog=23 | 25               | 28                                   | 27                                                                   |
| -16 wlog=23 | 25               | 26                                   | 25                                                                   |
| -13 wlog=22 | 24               | 25                                   | 24                                                                   |
| -11 wlog=22 | 24               | 25                                   | 25                                                                   |
| -11 wlog=27 | 29               | 25                                   | 25                                                                   |
| -9 wlog=21  | 23               | 23                                   | 23                                                                   |
| -7 wlog=21  | 23               | 23                                   | 23                                                                   |
| -3 wlog=21  | 23               | 21                                   | 21                                                                   |
| -3 wlog=27  | 29               | 21                                   | 21                                                                   |
| -1 wlog=19  | 21               | 21                                   | 21                                                                   |
| -1 wlog=22  | 25               | 21                                   | 21                                                                   |
| -1 wlog=27  | 29               | 21                                   | 21                                                                   |

### CODE REVIEW

**Comment by Cyan4973** [comment]
Thanks, this is a good parameter to re-balance.

2 metrics are important here : 
- Compare compression ratio between `dev` and the new formula (when they differ in job sizes)
   - we expect small compression ratio losses as job sizes get smaller, albeit very small ones
   - btlazy2 may be immune to this impact 
- Compare compression speed
   - for situations where job sizes are very large, reducing the job size will make it possible to have more threads in parallel. 
   - however, since it also increases the nb of jobs, in increases the cost of loading. Compare with "single worker" mode too (default), where more jobs may lead to slower speeds 
      - probably more perceptible for higher compression levels, especially btlazy2

Add this information to the table, it will be helpful to guide decision.
I'm leaning towards smaller job sizes, but we have to be aware of the impact.

**Comment by terrelln** [comment]
I think the job size for the optimal parser modes should probably be smaller.

In optimal parser modes we generally expect to be able to index between `2^(cyclceLog)` and `2^(cycleLog + 1)` indices. So I would maybe suggest `cycleLog + 2` or `cycleLog + 3`.

**Comment by senhuang42** [comment]
**jobLog**
| Conf          | --single-thread | chainLog + 4 | cycleLog+4 | cycleLog+3 | cycleLog+2 |
|---------------|----------------------|-------------------|------------|------------|------------|
| -22 wlog=27   | N/A                  | 30                | 30         | 29         | 28         |
| -19 wlog=27   | N/A                  | 28                | 27         | 26         | 25         |
| -19 wlog=23   | N/A                  | 28                | 27         | 26         | 25         |
| -18 wlog=23   | N/A                  | 27                | 26         | 25         | 24         |
| -17 wlog=23   | N/A                  | 27                | 26         | 25         | 24         |
| -16 wlog=23   | N/A                  | 26                | 25         | 24         | 23         |
| -15 wlog = 22 | N/A                  | 27                | 26         | 25         | 24         |
| -14 wlog = 22 | N/A                  | 26                | 25         | 24         | 23         |
| -13 wlog=22   | N/A                  | 25                | 24         | 23         | 22         |

**compressed size: silesia.tar (211950592)**
| Conf        | --single-thread | chainLog + 4 | cycleLog+4 | cycleLog+3 | cycleLog+2 |
|-------------|-----------------------------|----------------------------|------------|------------|------------|
| -22 wlog=27 | 52695566                    | 52695282                   | 52695282   | 52695282   | 52695282   |
| -19 wlog=27 | 53036064                    | 53032764                   | 53033624   | 53041095   | 53078593   |
| -19 wlog=23 | 53255527                    | 53258082                   | 53258131   | 53257754   | 53257118   |
| -18 wlog=23 | 53683661                    | 53681286                   | 53687891   | 53695021   | 53784884   |
| -17 wlog=23 | 54577628                    | 54578153                   | 54582370   | 54588802   | 54667284   |
| -16 wlog=23 | 55447771                    | 55454265                   | 55453816   | 55495538   | 55664545   |
| -15 wlog=22 | 57310659                    | 57315091                   | 57330585   | 57357453   | 57419607   |
| -14 wlog=22 | 57667305                    | 57685766                   | 57700342   | 57742959   | 57932455   |
| -13 wlog=22 | 58126529                    | 58146609                   | 58174742   | 58311605   | 58678517   |

**compression time (fastest of 3 runs): silesia.tar**
| Conf        |  --single-thread | chainLog + 4 | cycleLog+4   | cycleLog+3   | cycleLog+2   |
|-------------|-----------------------------|----------------------------|--------------|--------------|--------------|
| -22 wlog=27 | 1m:44s, 99%                 | 1m:47s, 99%                | 1m:46s, 99%  | 1m:45s, 99%  | 1m:45s, 99%  |
| -19 wlog=27 | 1m:21s, 99%                 | 1m:17s, 99%                | 56.22s, 171% | 33.70s, 315% | 22.24s, 545% |
| -19 wlog=23 | 1m:18s, 99%                 | 1m:13s, 99%                | 52.88s, 159% | 31.68s, 304% | 21.30s, 549% |
| -18 wlog=23 | 1m:02s, 99%                 | 44.60s, 150%               | 28.38s, 279% | 18.78s, 493% | 17.40s, 531% |
| -17 wlog=23 | 51.21s, 99%                 | 39.64s, 147%               | 22.95s, 287% | 15.61s, 509% | 14.25s, 537% |
| -16 wlog=23 | 41.00s, 99%                 | 20.14s, 293%               | 12.34s, 527% | 11.90s, 542% | 10.38s, 610% |
| -15 wlog=22 | 33.33s, 99%                 | 24.45s, 142%               | 15.504, 265% | 11.28s, 442% | 10.10s, 490% |
| -14 wlog=22 | 24.63s, 99%                 | 11.30s, 273%               | 8.60s, 442%  | 7.98s, 499%  | 7.63s, 522%  |
| -13 wlog=22 | 19.32s, 99%                 | 6.54s, 472%                | 6.51s, 508%  | 6.10s, 528%  | 5.40s, 614%  |

There's only a very small `cSize` regression (just a few KB) when using `cycleLog + 3` at the higher strategies, but smaller jobSizes are pretty significantly faster.

Based on these results, I'd say it would make sense to enable `cycleLog + 3` for `strategy >= btopt`. Otherwise, we use `cycleLog + 4`.

**Comment by senhuang42** [comment]
As a follow-up to also measure the rough cost of _loading_ the data, which might matter for cases with more jobs, I measured the speed of compression (total time) using `-T1 --long` with otherwise the same settings. I also included a run with `jobSize > srcSize` (512 MB for the 200MB `silesia.tar`) as another point of comparison.

**compression time: silesia.tar, fastest of 3 runs, -T1 --long**
| Conf        | -B512MB | chainLog + 4 | cycleLog+4 | cycleLog+3 | cycleLog+2 |
|-------------|------------------|--------------|------------|------------|------------|
| -22 wlog=27 | 1m:45s           | 1m:45s       | 1m:43s     | 1m:45s     | 1m:44s     |
| -19 wlog=27 | 81.42s           | 84.79s       | 86.42s     | 86.77s     | 85.71s     |
| -19 wlog=23 | 77.38s           | 79.90s       | 77.39s     | 80.83s     | 89.66s     |
| -18 wlog=23 | 64.82s           | 61.53s       | 59.36s     | 66.64s     | 71.21s     |
| -17 wlog=23 | 51.77s           | 51.73s       | 54.58s     | 57.27s     | 57.74s     |
| -16 wlog=23 | 42.27s           | 42.79s       | 43.64s     | 47.69s     | 48.32s     |
| -15 wlog=22 | 31.96s           | 31.63s       | 31.98s     | 33.82s     | 34.06s     |
| -14 wlog=22 | 23.55s           | 23.26s       | 23.76s     | 24.81s     | 23.60s     |
| -13 wlog=22 | 18.73s           | 19.17s       | 19.34s     | 19.53s     | 19.91s     |

Generally, it doesn't seem like the cost of loading seems to be too significant of a factor across the board, though there are slight, but noticeable slower compression times for some of the higher compression levels with smaller `jobSizes`. I don't think this changes the general conclusion though - `cycleLog + 3` seems to be a pretty good setting overall compared to `chainLog + 4`, and for a tiny penalty in cost of loading and compressed size, we get very big benefit in the user experience/speed of compressing with `-T0`.

**Comment by Cyan4973** [comment]
+3 it is then !

**Review by Cyan4973 — APPROVED** [review]

File: lib/compress/zstdmt_compress.c [hunk 1]
```diff
@@ -1184,8 +1184,8 @@ static unsigned ZSTDMT_computeTargetJobLog(const ZSTD_CCtx_params* params)
     if (params->ldmParams.enableLdm) {
         /* In Long Range Mode, the windowLog is typically oversized.
          * In which case, it's preferable to determine the jobSize
-         * based on chainLog instead. */
-        jobLog = MAX(21, params->cParams.chainLog + 4);
+         * based on cycleLog instead. */
+        jobLog = MAX(21, ZSTD_cycleLog(params->cParams.chainLog, params->cParams.strategy) + 4);
```
Comment (Cyan4973) [inline lib/compress/zstdmt_compress.c:1188]: `+ 3`

**Pull request merged**